### ISSUE TITLE
support rustls

### ISSUE THREAD

**Issue by rogerpengyu** [prisma/tiberius#202]
tiberius by default use OpenSSL. In one of our projects, we hit reliability issues with OpenSSL and found [rustls](https://docs.rs/rustls/latest/rustls/index.html) provides a better option for TLS.

To unblock myself, I create a fork to use rustls. Initially I want to add a feature to I could later merge back. But I found I have to change some interfaces from future::AsyncRead to tokio::AsyncRead. This makes the code really dirty. So I give adding as a feature and just adapt the code to tokio::AsyncRead.

I would love to merge back the change. But not sure what's the best option here. What's the suggestion?

**Comment by pimeys** [comment]
What, you got it to work with rustls on all platforms? This is great, makes Tiberius much better compared to the balancing we do with OpenSSL due to Security Framework not playing well with SQL Server.

To the `AsyncRead` question: please continue using the traits from the futures crate. Tokio has conversion traits between theirs and futures interfaces. You can find them from [tokio_util crate](https://docs.rs/tokio-util/0.7.1/tokio_util/compat/index.html).

I'd be very happy with rustls, and so would Prisma. It would cut lots of complexity if it'd give a working TLS on all platforms. One thing I'm not so sure about is certain enterprise users of Tiberius, that use Windows and expect to use schannel for TLS due to certain requirements. Would this change cut them out on using the crate? This is purely hypothetical and I don't know is there any such companies using this code internally.

P.S. There's already some previous work trying to use rustls, with not that great success: https://github.com/prisma/tiberius/pull/82

**Comment by rogerpengyu** [comment]
Hmm, the problem is I used another crate tokio-rustls to get a TlsStream implementation. It uses tokio::AsyncRead. I need to see how I should wrapper it so it doesn't look too ugly integrate with other parts. Give me a couple days, I will share the implementation to get concrete suggestions how to merge back.

By the way, do you have plan to support AAD authentication?

**Comment by pimeys** [comment]
We have users asking for AAD. It's not prioritized very high from our side currently, and requires some research from our part before writing any code to see how it works (I have no idea yet).

**Comment by rogerpengyu** [comment]
here is the pull request, please take a look: https://github.com/prisma/tiberius/pull/205

**Comment by rogerpengyu** [comment]
https://github.com/prisma/tiberius/pull/205 passed all gates. waiting for approval to be merged

**Comment by rogerpengyu** [comment]
Thanks for merging it. for my production, I am pointing to my private branch. When could I switch over?

**Comment by janpio** [comment]
Hey @rogerpengyu, we had an internal discussion about this today. As you probably figured,  @pimeys is pretty excited about this and really just wants to merge it. But there are others, that are a bit concerned about completely switching over Tiberius to this method as it changes its behavior quite a lot. 

Instead of being able to update OpenSSL via the system, the update now has to happen via an update of Tiberius and its dependencies. Depending on who you are talking to, this might be a benefit - or a disadvantage (use case: Someone that does not want to update the program using tiberius, but still needs to make sure to get security updates).

Does that explanation make sense to you?

One of the suggestions was that instead of _switching_, this new functionality is added via a _feature flag_, so the user of tiberius can chose the method themselves. Do you think this makes sense and would be doable with the code you started?

**Comment by rogerpengyu** [comment]
make sense. Yes I could put it under a feature flag. Might get some time within this week for that.

**Comment by pimeys** [comment]
Maybe a switch between native-tls and rustls.

**Comment by rogerpengyu** [comment]
based on the feedback, create a new PR: https://github.com/prisma/tiberius/pull/212

**Comment by pimeys** [comment]
Support in version 0.9.0. To use rustls instead of native-tls, compile without default features and with feature flag rustls.
